sound/soc/sof/ipc4-telemetry2.c

```diff
@@ -0,0 +1,81 @@
+
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry2.h"
+#include "ipc4-priv.h"
+
+static ssize_t sof_telemetry2_entry_read(struct file *file, char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	u32 type = SOF_IPC4_DEBUG_SLOT_TELEMETRY2;
+	loff_t pos = *ppos;
+	size_t size_ret;
+	u32 offset;
+	u8 *buf;
+
+	if (pos < 0)
+		return -EINVAL;
+	if (pos >= SOF_IPC4_DEBUG_SLOT_SIZE || !count)
+		return 0;
+	if (count > SOF_IPC4_DEBUG_SLOT_SIZE - pos)
+		count = SOF_IPC4_DEBUG_SLOT_SIZE - pos;
+
+	offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, type);
+	if (!offset)
+		return -EFAULT;
```

Comment:
the kernel should continue if it doesn't find an ID that only supported in newer versions of firmware.
I think this feature should be clearly advertised in the extended manifest, checking for the presence of an ID is a bit weird.

In this case it's not lethal because we actually ignore ALL errors from this function, so even returning -EFAULT is useless. You don't need an error code...